packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -0,0 +1,563 @@
+use std::fmt::Display;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use cosmrs::Any;
+use error_stack::{report, Context, Report, ResultExt};
+use events::{AbciEventTypeFilter, Event};
+use futures::{pin_mut, Stream, TryFutureExt};
+use itertools::Itertools;
+use mockall::automock;
+use report::LoggableError;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::time::interval;
+use tokio_stream::{Elapsed, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{error, info};
+use typed_builder::TypedBuilder;
+use valuable::Valuable;
+
+use crate::grpc::client;
+use crate::utils::{with_retry, RetryPolicy};
+
+type StreamElement = Result<error_stack::Result<Event, client::Error>, Elapsed>;
+
+#[automock(
+    type Err = Error;
+    type Event = Event;
+)]
+#[async_trait]
+pub trait EventHandler: Send + Sync {
+    type Err: Context;
+    type Event: TryFrom<Event>;
+
+    async fn handle(
+        &self,
+        event: &Self::Event,
+        token: &CancellationToken,
```

Comment:
Thought we decided to wrap this in the `Context` struct?

Reply:
true, I forgot about that